rfc/halyard-lite.md

```diff
@@ -0,0 +1,473 @@
+as Halyard generates and applies YAML, there is no intermediate representation
+that users can customize using these tools.
+
+While the Armory operator provides a viable Kubernetes-native way to install Spinnaker,
+there is still no Kubernetes-native install path for users who won’t want or need
+an operator.  Ideally it would be reasonably easy to write a Helm chart and/or
+Kustomization that installs Spinnaker’s microservices and is managed using idiomatic
+patterns.
+
+### Complexity of implementation
+
+Halyard is the third-largest of Spinnaker’s microservices by lines of code, comprising
+55435 lines of Java code.  There is minimal test code, totalling 3094 lines of Java and
+Groovy code.  The poor test coverage, combined with heavy use of reflection and unchecked
+casts make refactoring difficult.
+
+Halyard also consumes many of the other microservices as libraries.  In general,
+engineers modifying the other microservices don’t expect that they are modifying
+a library consumed by other microservices.  This has led to cases where changes
+to clouddriver or front50 led to [breaking Halyard](https://github.com/spinnaker/halyard/pull/1062)
+when trying to bump the dependency.
+
+As a representative example of the amount of code change required to add a new
+config parameter to Halyard, consider the PR to add [Google Pubsub config](https://github.com/spinnaker/halyard/pull/750).
+This is a fairly representative example of a config parameter; it supports add/edit/delete/get/list
+on a list of Google Pubsub subscriptions, and required 2244 lines of code to
+implement (none of which is test code).
+
+The amount of code necessary to add simple config parameters is a burden both on
+contributors and on the maintainers of Halyard who have to review this code.  This
+has led to Halyard’s config often lagging behind that of the individual microservices.
+Ideally, adding a config parameter would be a reasonably simple first contribution
+for someone looking to get involved with the project, but that is currently very
+far from the case.
+
+## Timeline
+
+* Spinnaker 1.19 (March 2020)
+  * An alpha version of Halyard Lite is released and is able to generate service
+  configurations from the Halyard config.
+  * A simple Kustomize kustomization is released that can deploy Spinnaker to
+  Kubernetes in simple cases.
+* Spinnaker 1.20 (May 2020)
+  * Feedback from early users of Halyard Lite and the kustomization are
+   addressed, and the kustomization now supports more advanced use cases (ex:
+   high availability, custom BOM, etc.)
+  * The Armory operator is moved to depend on Halyard Lite
+* Spinnaker 1.21 (July 2020)
+  * Halyard Lite, as well as the kustomize install path are GA and the
+   recommended install path for new users of Spinnaker.
+  * Based on adoption of Halyard Lite, a plan is formulated around how and when
+  to deprecate and cease support of Halyard. The details of this plan will be a
+  separate RFC.
```

Comment:
I assume this also includes plans for updating the documentation?

Reply:
Yes, good point!  There will of course be documentation on Halyard Lite accompanying the alpha release so that early adopters can use it---but as part of the deprecation of Halyard we'll need to update all of the documentation that references `hal` commands to instead show how to update the halyard config.  I'll update the RFC to make this explicit.